### PR TITLE
Discovery: Implement focus handling for color picker

### PR DESCRIPTION
**Pull request by illia-stv**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Internal(color-picker): Implement focus handling for color picker.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-font/tests/ui/colortableview.js [hunk 1]
```diff
@@ -616,6 +616,20 @@ describe( 'ColorTableView', () => {
...
+				stopPropagation: sinon.spy()
+			};
+
+			dropdown.colorTableView.appendColorPicker();
+
+			dropdown.colorTableView.focusTracker.isFocused = true;
+
+			dropdown.colorTableView.keystrokes.press( keyEvtData );
+			sinon.assert.calledOnce( keyEvtData.stopPropagation );
+		} );
```
Comment (Dumluregn) [inline packages/ckeditor5-font/tests/ui/colortableview.js:632]: Putting this test inside `describe( 'updateSelectedColors() with document colors', ()` _`=>`_ `{` block seems weird, it should rather be in the `describe( 'appendColorPicker()', ()` _`=>`_ `{` one.

**Review by Dumluregn — CHANGES_REQUESTED** [review]
PR works, and it also closes another issue (keyboard handling), so good job :+1: Let's polish it a bit. Besides a couple of small refactors, I miss the integration unit test for focus cycler - meaning that consecutively pressing `tab` and `shift+tab` we move to the expected parts of the component. So the logic in tests should be:  
1) focus element X (e.g. last document color tile)

2) press `tab`

3) check if the element Y (here - first color picker slider) is focused.

etc. for other components. It's already tested partially [here](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-font/tests/ui/colortableview.js#L174-L242), but without the color picker parts (as `appendColorPicker()` is not called in the test), so let's enhance this test suite.

**Comment by Dumluregn** [comment]
One more remark - let's add a clearer indication that slider is focused. We discussed it on Slack and this is the option we want to go with - [https://cksource.slack.com/archives/C014DN0KB1S/p1680611929301089?thread_ts=1680608490.373359&cid=C014DN0KB1S,](https://cksource.slack.com/archives/C014DN0KB1S/p1680611929301089?thread_ts=1680608490.373359&cid=C014DN0KB1S,) so please apply this style on the focused slider.

File: packages/ckeditor5-ui/src/colorpicker/colorpickerview.ts [hunk 2]
```diff
@@ -75,6 +81,18 @@ export default class ColorPickerView extends View {
 
 		this.picker = global.document.createElement( 'hex-color-picker' );
 		this.picker.setAttribute( 'class', 'hex-color-picker' );
+		this.picker.setAttribute( 'tabindex', '-1' );
+
+		const colorPickersChildren = [ ...this.picker.shadowRoot!.children ] as Array<HTMLElement>;
+		const sliders = colorPickersChildren.filter( item => item.tagName === 'DIV' );
```
[the remark this answers is not in the document]
Comment (illia-stv) [reply packages/ckeditor5-ui/src/colorpicker/colorpickerview.ts:87]: I thought about this solution but unfortunately in our HTMLElement types there is no **#role** property, so I decided to filter by div's I think its safe, becouse we freezed package version of color picker library, so nothing will be changed there.  
image

**Comment by illia-stv** [comment]
> One more remark - let's add a clearer indication that slider is focused. We discussed it on Slack and this is the option we want to go with - https://cksource.slack.com/archives/C014DN0KB1S/p1680611929301089?thread_ts=1680608490.373359&cid=C014DN0KB1S, so please apply this style on the focused slider.

It's a little bit tricky to style it due vanilla-colorful uses shadow dom under the hood. Let's move it to followup: https://github.com/cksource/ckeditor5-internal/issues/3050

File: packages/ckeditor5-ui/src/colorpicker/colorpickerview.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pszczesniak) [reply packages/ckeditor5-ui/src/colorpicker/colorpickerview.ts:87]: Hmmm :thinking: looks fine in my `VSCode` - try one more time, do `yarn reinstall` or reopen the `VSCode`.  
 
Screenshot 2023-04-06 at 14 09 04

**Comment by pszczesniak** [comment]
I've push a [small update](https://github.com/ckeditor/ckeditor5/commit/702ae5401bbaa79daea9c9add3ca956ec4ab7fef) of how messages that need localization should be passed, based on [documentation](https://ckeditor.com/docs/ckeditor5/latest/framework/deep-dive/ui/localization.html#writing-a-localizable-ui).

**Comment by pszczesniak** [comment]
Followup cksource/ckeditor5-internal#3065

**Review by pszczesniak — CHANGES_REQUESTED** [review]
I'm waiting for fixes that were reported by @Dumluregn and marked by You as resolved to look at it one more time.

File: packages/ckeditor5-font/src/ui/colortableview.ts [hunk 3]
```diff
@@ -317,6 +318,29 @@ export default class ColorTableView extends View {
 		} );
 
 		this.items.add( this.colorPickerView );
+
+		this._focusColorPickersElements();
+
+		this._stopPropagationOnArrowsKeys();
+	}
+
+	private _focusColorPickersElements() {
```
Comment (pszczesniak) [inline packages/ckeditor5-font/src/ui/colortableview.ts:327]: Missing API docs and `type` of what this function returns.

File: packages/ckeditor5-font/src/ui/colortableview.ts [hunk 4]
```diff
@@ -317,6 +318,29 @@ export default class ColorTableView extends View {
...
+		for ( const slider of this.colorPickerView!.slidersView ) {
+			this.focusTracker.add( slider.element! );
+			this._focusables.add( slider );
+		}
+
+		this.focusTracker.add( this.colorPickerView!.input.element! );
+		this._focusables.add( this.colorPickerView!.input );
+	}
+
+	private _stopPropagationOnArrowsKeys() {
```
Comment (pszczesniak) [inline packages/ckeditor5-font/src/ui/colortableview.ts:337]: Missing API docs and `type` of what this function returns.

**Comment by pszczesniak** [comment]
CI is failing - some tests are not passing.

**Review by pszczesniak — APPROVED** [review]
LGTM :+1:

**Comment by mlewand** [comment]
There's a unit test that is skipped because main branch doesn't have the changes of this PR: [https://github.com/ckeditor/ckeditor5/blob/3efbea82164a77f387fd614486a01eed24a70a18/packages/ckeditor5-ui/tests/colorpicker/colorpickerview.js#L66](https://github.com/ckeditor/ckeditor5/blob/3efbea82164a77f387fd614486a01eed24a70a18/packages/ckeditor5-ui/tests/colorpicker/colorpickerview.js#L66)

Please remove the `skip` in your PR - this should give a full CC.

(more in https://github.com/ckeditor/ckeditor5/pull/13820#issuecomment-1502930260)

**Comment by illia-stv** [comment]
> There's a unit test that is skipped because main branch doesn't have the changes of this PR:
> 
> https://github.com/ckeditor/ckeditor5/blob/3efbea82164a77f387fd614486a01eed24a70a18/packages/ckeditor5-ui/tests/colorpicker/colorpickerview.js#L66
> 
> Please remove the `skip` in your PR - this should give a full CC.
> 
> (more in [#13820 (comment)](https://github.com/ckeditor/ckeditor5/pull/13820#issuecomment-1502930260))

After Piotrek's last merge, this **skip** was magically removed, look: https://github.com/ckeditor/ckeditor5/blob/4896e4d9c4eac89c95f94dd3deb1c42e9e83db7f/packages/ckeditor5-ui/tests/colorpicker/colorpickerview.js#L66

**Pull request merged**